File: content/05/code/ruby/20-implement-maximum-length/lib/shouty.rb
module Shouty
  class Person
    attr_reader :messages_heard, :location

    def initialize(network, location)
      @messages_heard = []
      @network        = network
      @location       = location

      @network.subscribe(self)
    end

    def shout(message)
      @network.broadcast(message, @location)
    end

    def hear(message)
      @messages_heard << message
    end
  end

  class Network
    def initialize(range)
      @range     = range
      @listeners = []
    end

    def subscribe(person)
      @listeners << person
    end

    def broadcast(message, shouter_location)
      @listeners.each do |listener|
        if (listener.location - shouter_location).abs <= @range
          if message.size <= 180
            listener.hear(message)
          end
        end
      end
    end
  end
end

File: content/bdd-with-cucumber/05/code/js/05-location-specified-in-data-table/src/shouty.js
class Person {
  constructor(network, location) {
    this.messages = []
    this.network  = network
    this.location = location


    this.network.subscribe(this)
  }

  shout(message) {
    this.network.broadcast(message, this.location)
  }

  hear(message) {
    this.messages.push(message)
  }

  messagesHeard() {
    return this.messages
  }
}

class Network {
  constructor(range) {
    this.listeners = []
    this.range     = range
  }

  subscribe(person) {
    this.listeners.push(person)
  }

  broadcast(message, shouter_location) {
    this.listeners.forEach(listener => {
      if( Math.abs(listener.location - shouter_location) <= this.range)
        listener.hear(message)
    })
  }
}

module.exports = {
  Person  : Person,
  Network : Network
}
